### PR TITLE
Food Module

### PR DESCRIPTION
**Pull request by jakkeeny**
This is a modified, further developed and optimised version of the CSE food rations module.

Adds
- 2x MRE Item Types
- 3x Waterbottle Item Types
- Camelbak Item
- Food and Drink Consumption
- Food and Drink Reduction

### CODE REVIEW

File: addons/food/functions/fnc_consume.sqf [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+#include "../script_component.hpp"
+
+private ["_player", "_mode"];
+
+_player = _this select 1;
+_mode = _this select 2;
+
+switch (_mode) do {
+    case "eat": {
```
Comment (dedmen) [inline addons/food/functions/fnc_consume.sqf]: Maybe better make mode a number enum?

File: addons/food/functions/fnc_consume.sqf [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/*
+	File Name: fnc_consume.sqf
+	Author: Jak Keen
+*/
+
+#include "../script_component.hpp"
+
+private ["_player", "_mode"];
+
+_player = _this select 1;
```
Comment (dedmen) [inline addons/food/functions/fnc_consume.sqf]: params

File: addons/food/functions/fnc_consume.sqf [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    case "eat": {
+        private ["_mre"];
+
+        _mre == "";
+
+        {
+            if(_x in (items player)) exitwith {
+                _mre = _x;
+            };
+        } forEach GVAR(food_items);
```
Comment (dedmen) [inline addons/food/functions/fnc_consume.sqf]: you are doing this everywhere.
you can also use the CBA inventoryChanged eventhandler to only check that once on inventory change. Without the inventory changing the contents of the players inventory cannot change anyway.

File: addons/food/functions/fnc_consume.sqf [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+        if(!(_mre == "")) then {
```
Comment (dedmen) [inline addons/food/functions/fnc_consume.sqf]: if (not is equal)? How about if (is not equal)?
`if (_mre != "")`

File: addons/food/functions/fnc_consume.sqf [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            };
+        } forEach GVAR(drink_items);
+
+        if(_drink == "") exitWith {
+            hint "You need to have a drink on you before you can drink";
+        };
+
+        player say3D "drinking";
+        _player removeItem _drink;
+        _item = GET_CFG_VAR("CfgWeapons", _drink, "ace_onDrink");
```
Comment (dedmen) [inline addons/food/functions/fnc_consume.sqf]: honestly I prefer that written out. You aren't saving much here anyway

File: addons/food/functions/fnc_hasCamelbak.sqf [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+#include "../script_component.hpp"
+
+private ["_item", "_return"];
+
+_item = "";
+_return = false;
+
+if(GVAR(camelbak_item) in (items player)) exitwith {
+    _item = _x;
```
Comment (dedmen) [inline addons/food/functions/fnc_hasCamelbak.sqf]: where does the ? `_x` come from?

Comment (dedmen) [reply addons/food/functions/fnc_hasCamelbak.sqf]: Seeing that this is called from https://github.com/acemod/ACE3/pull/6339/files#diff-ca5c2e394d87bed18d3bae861f257a91R23 which is not in a forEach loop. This is a script error.

File: addons/food/functions/fnc_reduceStats.sqf [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+if([] call FUNC(hasCamelbak)) then {
+    _amount = 100 - GVAR(water);
+    GVAR(camelbak) = GVAR(camelbak) - _amount;
+    GVAR(water) = water + _amount;
+};
+
+if (GVAR(water) < 20 || GVAR(food) < 20) then {
+    if (((speed _player) > 12) && ((vehicle _player) == _player)) then {
+        _player playMove "amovppnemstpsraswrfldnon";
```
Comment (dedmen) [inline addons/food/functions/fnc_reduceStats.sqf]: not MP synchronized I think.

File: addons/food/functions/fnc_refill.sqf [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+private ["_item", "_player", "_mode"];
+
+_player = _this select 1;
+_mode = _this select 2;
+
+switch (_mode) do {
+    case "bottle": {
+        {
+            if (_x in (items player)) exitwith {
+                if(count (nearestObjects [player, GVAR(refill_sources), 2]) > 0) then {
```
Comment (dedmen) [inline addons/food/functions/fnc_refill.sqf]: The refill action is only available if you have an item to refill and if there is a refill source nearby. Why are you checking that again here?

File: addons/food/functions/fnc_refill.sqf [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+                    player addItem _item;
+                };
+            };
+        } forEach GVAR(refill_items);
+    };
+
+    case "camelbak": {
+        {
+            if (_x in (items player)) exitwith {
+                if(GVAR(camelbak_item) in (items player)) then {
```
Comment (dedmen) [inline addons/food/functions/fnc_refill.sqf]: same as above

File: addons/food/functions/fnc_updatePlayerHud.sqf [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+*/
+
+#include "../script_component.hpp"
+
+private ["_ui", "_ctrlStatusFood", "_ctrlStatusDrink", "_ctrlStatusCamelbak", "_drinkStatus", "_foodStatus", "_camelbakStatus", "_foodColour", "_drinkColour", "_camelbakColour"];
+
+disableSerialization;
+_ui = uiNamespace getVariable "ACE_FoodStats";
+
+if (!isNil "_ui") then {
```
Comment (dedmen) [inline addons/food/functions/fnc_updatePlayerHud.sqf]: exitWith.

File: addons/food/functions/fnc_updatePlayerHud.sqf [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    if (_camelbakStatus <= 32) then {
+        _camelbakColour = [1, 0, 0, 1];
+    };
+
+    if (_camelbakStatus >= 33) then {
+        _camelbakColour = [1, 0.576, 0.059, 1];
+    };
+
+    if (_camelbakStatus >= 66) then {
+        _camelbakColour = [0, 1, 0, 1];
```
Comment (dedmen) [inline addons/food/functions/fnc_updatePlayerHud.sqf]: How about just linear interpolate and get the full color range even easier and faster than this?

File: addons/food/script_component.hpp [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    #define DEBUG_MODE_FULL
+#endif
+
+#ifdef DEBUG_ENABLED_FOOD
+    #define DEBUG_SETTINGS DEBUG_SETTINGS_FOOD
+#endif
+
+#include "\z\ace\addons\main\script_macros.hpp"
+
+#define GET_CFG_VAR(CFG, ITEM, VAR) (getText (configFile >> CFG >> ITEM >> VAR))
```
Comment (dedmen) [inline addons/food/script_component.hpp]: `GET_CFG_VAR` -> Get config Variable. It's unclear that it returns text.

File: addons/food/README.md [hunk 13]
```diff
@@ -0,0 +1,11 @@
+ace_food
+========
+
+Custom fonts including fixed-width font.
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/README.md]: Copied the readme from ace_fonts, is an error and I will correct

File: addons/food/functions/fnc_refill.sqf (hunk 8 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_refill.sqf]: Fair point, I will remove

File: addons/food/functions/fnc_addActions.sqf [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        _con;
+    };
+
+    _conRefill = {
+        private ["_con", "_sources"];
+        _con = false;
+
+        {
+            if (_x in (items player)) exitwith {
+                if(count (nearestObjects [player, GVAR(refill_sources), 2]) > 0) then {
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/functions/fnc_addActions.sqf]: What way would you recommend doing it?

File: addons/food/functions/fnc_consume.sqf (hunk 1 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_consume.sqf]: Yes, I will do that

File: addons/food/functions/fnc_consume.sqf (hunk 2 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_consume.sqf]: Will do

File: addons/food/functions/fnc_consume.sqf (hunk 4 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_consume.sqf]: Done

File: addons/food/functions/fnc_hasCamelbak.sqf (hunk 6 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_hasCamelbak.sqf]: This is an error and has now been rewritten

File: addons/food/functions/fnc_consume.sqf (hunk 5 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_consume.sqf]: Done

File: addons/food/functions/fnc_refill.sqf (hunk 9 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_refill.sqf]: Done

File: addons/food/functions/fnc_updatePlayerHud.sqf (hunk 10 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_updatePlayerHud.sqf]: Done

File: addons/food/script_component.hpp (hunk 12 above)
Comment (jakkeeny) [reply addons/food/script_component.hpp]: Removed

File: addons/food/CfgWeapons.hpp [hunk 15]
```diff
@@ -0,0 +1,125 @@
+class CfgWeapons {
+    class Default;
+    class InventoryItem_Base_F;
+    
+    class ItemCore: Default {
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/CfgWeapons.hpp]: Changed

File: addons/food/CfgWeapons.hpp [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+            disabled = 0;
+        };
+        class ItemInfo: InventoryItem_Base_F {
+            mass= 5;
+            type= 201;
+        };
+    };
+
+    // Camelbaks
+    class ACE_Camelbak: ItemCore {
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/CfgWeapons.hpp]: Changed

File: addons/food/CfgWeapons.hpp [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+        displayName = "Camelbak";
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/CfgWeapons.hpp]: Done

**Comment by jokoho48** [comment]
there is already a port of CSE Field Radions in ACEX https://github.com/acemod/ACEX/pull/114

File: addons/food/CfgWeapons.hpp [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+    };*/
+
+    // Waterbottles
+    class ACE_Waterbottle_Full: ItemCore {
+        displayName = "Water Bottle (Full)";
+        picture = QPATHTOF(data\icons\waterbottle_full.paa);
+        model = "\A3\Structures_F_EPA\Items\Food\BottlePlastic_V2_F.p3d";
+        descriptionShort = "A full waterbottle";
+
+        ace_onDrink = "ACE_Waterbottle_Part";
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/CfgWeapons.hpp]: Done

File: addons/food/XEH_postInit.sqf [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+// Add eat/drink actions
+[] call FUNC(addActions);
+
+// Init reduction of stats
+["itemAdd", ["ACE_FoodStats_Loop", {
+    if (alive player) then {
+        [player] call FUNC(reduceStats);
+    };
+}, 4]] call BIS_fnc_loop;
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/XEH_postInit.sqf]: Done

File: addons/food/XEH_postInit.sqf [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+
+// Call update loop file
+[{call FUNC(updatePlayerHud)}, 1, []] call CBA_fnc_addPerFrameHandler;
+
+// Add eat/drink actions
+[] call FUNC(addActions);
+
+// Init reduction of stats
+["itemAdd", ["ACE_FoodStats_Loop", {
+    if (alive player) then {
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/XEH_postInit.sqf]: I don't believe it to be

File: addons/food/config.cpp [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+                h = 0.09;
+            };
+        };
+    };
+};
+
+class CfgSounds {
+    sounds[] = {};
+
+    class eating {
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/config.cpp]: Done the ace prefix

File: addons/food/functions/fnc_addActions.sqf [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+	Author: Jak Keen
+*/
+
+#include "../script_component.hpp"
+
+private ["_condition", "_children", "_action"];
+
+_condition = {
+    private ["_con"];
+    _con = false;
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/functions/fnc_addActions.sqf]: done

File: addons/food/functions/fnc_addActions.sqf [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+
+    {
+        if (_x in (items player)) exitwith {
+            _con = true;
+        };
+    } forEach GVAR(food_items) + GVAR(drink_items) + GVAR(refill_items);
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/functions/fnc_addActions.sqf]: Used your superior solution

File: addons/food/functions/fnc_addActions.sqf [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+
+    _con;
+};
+
+_children = {
+    private ["_target", "_player", "_params", "_actions", "_action", "_condition", "_conDrink", "_conEat", "_conRefill", "_conCamelRefill"];
+    params ["_target", "_player", "_params"];
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/functions/fnc_addActions.sqf]: didn't know params privitized as well, fixed

File: addons/food/functions/fnc_consume.sqf (hunk 3 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_consume.sqf]: Couldn't find it originally so thought it didn't exist. Now changed

File: addons/food/functions/fnc_addActions.sqf (hunk 14 above)
[the remark this answers is not in the document]
Comment (dedmen) [reply addons/food/functions/fnc_addActions.sqf]: You could store the players position. And only refresh if he moved a meter or so.

**Comment by dedmen** [comment]
The ACEX version does everything this does. And has all the issues I commented fixed/implemented.

File: addons/food/functions/fnc_updatePlayerHud.sqf (hunk 11 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_updatePlayerHud.sqf]: Done

File: addons/food/functions/fnc_onCloseInventory.sqf [hunk 25]
```diff
@@ -0,0 +1,17 @@
+/*
+	File Name: fnc_onCloseInventory.sqf
+	Author: Jak Keen
+*/
+
+#include "../script_component.hpp"
+
+private ["_camelbaks"];
+
+if((items player) isEqualTo GVAR(inventory)) exitWith {};
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/functions/fnc_onCloseInventory.sqf]: Fixed

File: addons/food/XEH_postInit.sqf [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+
+// Set vars
+GVAR(food) = 100;
+GVAR(water) = 100;
+GVAR(camelbak) = 0;
+
+GVAR(food_decent_rate) = 0.02;
+GVAR(water_decent_rate) = 0.025;
+
+GVAR(food_items) = ["ACE_MRE_Type1", "ACE_MRE_Type2"];
```
[the remark this answers is not in the document]
Comment (jakkeeny) [reply addons/food/XEH_postInit.sqf]: Done

File: addons/food/functions/fnc_reduceStats.sqf (hunk 7 above)
Comment (jakkeeny) [reply addons/food/functions/fnc_reduceStats.sqf]: Should be MP synchronised

**Comment by alganthe** [comment]
This feature has already been merged in ACEX, closing.

**Pull request closed without merging**